File: cpp/tests/dump_whir.cpp
// ===========================================================================
// dump_whir.cpp — WHIR 协议 golden test (跨语言对拍基准, 多 case)
//
// 输入:  确定性 RNG (seed=0xCAFE), 每个 case 独立的协议参数
// 输出:  标准输出 — 每个 case 以 "CASE N" 分隔, 含协议各阶段中间值和 proof
//        与 Rust 端 examples/dump_whir.rs 输出格式完全一致, 可直接 diff
//
// 运行:
//   ./dump_whir > golden_whir_cpp.txt
//   diff golden_whir_rs.txt golden_whir_cpp.txt
//
// Case 覆盖:
//   CASE 0: 1 多项式 × 2 系数  (最小可用)
//   CASE 1: 1 多项式 × 4 系数
//   CASE 2: 1 多项式 × 8 系数
//   CASE 3: 3 多项式 × 4 系数  (多向量)
//   CASE 4: 4 多项式 × 8 系数  (多向量, 大参数)
// ===========================================================================

#include "whir/algebra/goldilocks.hpp"
#include "whir/algebra/embedding.hpp"
#include "whir/algebra/linear_form.hpp"
#include "whir/algebra/ntt/cooley_tukey_goldilocks.hpp"
#include "whir/algebra/utilities.hpp"
#include "whir/hash/sha2_engine.hpp"
#include "whir/protocols/irs_commit.hpp"
#include "whir/protocols/matrix_commit.hpp"
#include "whir/protocols/merkle_tree.hpp"
#include "whir/protocols/proof_of_work.hpp"
#include "whir/protocols/sumcheck_protocol.hpp"
#include "whir/protocols/whir/whir.hpp"
#include "whir/transcript/transcript.hpp"

#include <cstdint>
#include <cstdio>
#include <string>
#include <vector>

using F = ::whir::algebra::Goldilocks;
using Emb = ::whir::algebra::Identity<F>;

// ---- 确定性 RNG (与 Rust Lcg 完全一致) ----
struct Lcg {
    uint64_t s;
    explicit Lcg(uint64_t seed) : s(seed) {}
    uint64_t next() {
        s = s * 6364136223846793005ULL + 1442695040888963407ULL;
        return s;
    }
};

// ---- 输出辅助函数 ----

void dump_bytes(const char* label, const auto& data) {
    std::printf("    %s ", label);
    for (auto b : data) std::printf("%02x", static_cast<unsigned>(b));
    std::printf("\n");
}

void dump_field_vec(const char* label, const std::vector<F>& vec) {
    std::printf("    %s", label);
    for (auto& v : vec) std::printf(" %llu", (unsigned long long)v.as_canonical_u64());
    std::printf("\n");
}

// ---- DomainSeparator 构造 (所有 case 共用) ----
::whir::transcript::DomainSeparator make_ds(const char* session_label) {
    ::whir::transcript::DomainSeparator ds;
    uint8_t cbor_proto[] = {0x19, 0xBE, 0xEF};
    sha3_512_hash(cbor_proto, 3, ds.protocol_id.data());
    // session = CBOR("whir_") + label, 区分不同 case 的海绵状态
    std::string label_str = std::string("whir_") + session_label;
    std::vector<uint8_t> cbor_sess;
    cbor_sess.push_back(0x60 + static_cast<uint8_t>(label_str.size())); // CBOR text header
    for (char c : label_str) cbor_sess.push_back(static_cast<uint8_t>(c));
    sha3_256_hash(cbor_sess.data(), cbor_sess.size(), ds.session_id.data());
    return ds;
}

// ---- 单个 WHIR 测试 case ----
// 参数:
//   case_id      — 用于输出标识和 session 区分
//   num_vectors  — 多项式数量
//   vector_size  — 每个多项式的系数个数 (必须是 2 的幂)
//   rng          — 确定性随机数生成器 (被推进)
// 输出: 标准输出, 格式 "CASE N" + 协议各阶段值
void run_case(int case_id, std::size_t num_vectors, std::size_t vector_size, Lcg& rng) {
    std::size_t depth = 1;
    std::size_t cw = vector_size * 2;       // codeword = 2 * vector_size
    std::size_t num_cols = num_vectors * depth;
    std::size_t ood = 1;                    // 域外采样数
    std::size_t ind = std::min(vector_size, std::size_t{3}); // 域内采样 (≤ 向量大小)
    std::size_t log_vs = 0;
    { std::size_t sz = vector_size; while (sz > 1) { sz >>= 1; ++log_vs; } }

    std::printf("CASE %d\n", case_id);

    // ---- 1. 输入构造 ----
    // 每个多项式生成随机系数
    std::vector<std::vector<F>> vecs(num_vectors);
    std::vector<std::span<const F>> vec_spans;
    std::vector<F> evaluations;

    for (std::size_t vi = 0; vi < num_vectors; ++vi) {
        vecs[vi].resize(vector_size);
        for (auto& v : vecs[vi]) v = F::from_u64(rng.next());
        vec_spans.push_back(vecs[vi]);
        // 全 1 线性形式在每向量上的求值 = 系数之和
        F eval = F::zero();
        for (auto& v : vecs[vi]) eval += v;
        evaluations.push_back(eval);
        dump_field_vec(("input_vector" + std::to_string(vi)).c_str(), vecs[vi]);
    }

    // 一个全 1 线性形式 (对所有多项式通用)
    auto lf = std::make_unique<::whir::algebra::Covector<F>>(
        std::vector<F>(vector_size, F::one()));
    std::vector<std::unique_ptr<::whir::algebra::LinearForm<F>>> linear_forms;
    linear_forms.push_back(std::move(lf));

    // ---- 2. 协议配置 ----
    ::whir::protocols::irs_commit::Config<Emb> initial_committer;
    initial_committer.num_vectors = num_vectors;
    initial_committer.vector_size = vector_size;
    initial_committer.codeword_length = cw;
    initial_committer.interleaving_depth = depth;
    initial_committer.matrix_commit_num_cols = num_cols;
    initial_committer.in_domain_samples = ind;
    initial_committer.out_domain_samples = ood;
    initial_committer.deduplicate_in_domain = false;
    initial_committer.matrix_commit_mt = ::whir::protocols::merkle_tree::make_config(
        ::whir::hash::ENGINE_ID_SHA2, cw);

    ::whir::protocols::sumcheck::Config<F> initial_sumcheck;
    initial_sumcheck.initial_size = vector_size;
    initial_sumcheck.num_rounds = 0;

    ::whir::protocols::sumcheck::Config<F> final_sumcheck;
    final_sumcheck.initial_size = vector_size;
    final_sumcheck.num_rounds = log_vs;  // fold: vector_size → 1

    ::whir::protocols::whir::Config<Emb> whir_config;
    whir_config.initial_committer = initial_committer;
    whir_config.initial_sumcheck = initial_sumcheck;
    whir_config.initial_skip_pow = {};
    whir_config.final_sumcheck = final_sumcheck;
    whir_config.final_pow = {};

    // ---- 3. DomainSeparator (每个 case 独立 session) ----
    auto ds = make_ds(std::to_string(case_id).c_str());
    ::whir::transcript::Empty instance;

    // ---- 4. Prover ----
    auto ps = ::whir::transcript::ProverState::from_ds(ds, instance);
    auto witness = whir_config.commit(ps, vec_spans);

    dump_field_vec("witness_matrix", witness.matrix);
    dump_field_vec("ood_points", witness.out_of_domain.points);
    dump_field_vec("ood_matrix", witness.out_of_domain.matrix);

    std::vector<::whir::protocols::irs_commit::Witness<F, F>> whir_wits;
    whir_wits.push_back(std::move(witness));
    // 重新构造 spans (避免与 commit 内部移动冲突)
    std::vector<std::span<const F>> sp2;
    for (auto& v : vecs) sp2.push_back(v);

    auto claim = whir_config.prove(ps, sp2,
        std::span<const ::whir::protocols::irs_commit::Witness<F, F>>{whir_wits},
        std::move(linear_forms), evaluations);
    auto proof = std::move(ps).proof();

    dump_field_vec("eval_point", claim.evaluation_point);
    dump_field_vec("rlc_coeffs", claim.rlc_coefficients);
    dump_bytes("proof_narg", proof.narg_string);
    dump_bytes("proof_hints", proof.hints);

    // ---- 5. Verifier ----
    auto vs = ::whir::transcript::VerifierState::from_ds(ds, instance, proof);
    auto comm = whir_config.receive_commitment(vs);
    std::vector<const ::whir::protocols::irs_commit::Commitment<F>*> cp{&comm};
    auto vc = whir_config.verify(vs, cp, evaluations);

    dump_field_vec("v_eval_point", vc.evaluation_point);
    dump_field_vec("v_rlc_coeffs", vc.rlc_coefficients);
    std::printf("    check_eof %d\n", static_cast<int>(vs.check_eof()));
}

int main() {
    std::printf("# SECTION whir\n");

    Lcg rng(0xCAFE);

    // CASE 0: 1 多项式 × 2 系数 (最小)
    run_case(0, 1, 2, rng);

    // CASE 1: 1 多项式 × 4 系数
    run_case(1, 1, 4, rng);

    // CASE 2: 1 多项式 × 8 系数
    run_case(2, 1, 8, rng);

    // CASE 3: 3 多项式 × 4 系数 (多向量)
    run_case(3, 3, 4, rng);

    // CASE 4: 4 多项式 × 8 系数 (多向量 + 大参数)
    run_case(4, 4, 8, rng);

    return 0;
}

File: examples/dump_whir.rs
// ===========================================================================
// dump_whir.rs — WHIR 协议 golden test (跨语言对拍基准, 多 case)
//
// 运行: cargo run --example dump_whir --release > golden_whir_rs.txt
// 对拍: diff golden_whir_rs.txt golden_whir_cpp.txt
//
// Case 覆盖:
//   CASE 0: 1 多项式 × 2 系数  (最小可用)
//   CASE 1: 1 多项式 × 4 系数
//   CASE 2: 1 多项式 × 8 系数
//   CASE 3: 3 多项式 × 4 系数  (多向量)
//   CASE 4: 4 多项式 × 8 系数  (多向量, 大参数)
// ===========================================================================

use std::borrow::Cow;

use ark_ff::Field;
use ordered_float::OrderedFloat;

use whir::algebra::{
    embedding::Identity,
    fields::Field64,
    linear_form::{Covector, LinearForm},
};
use whir::hash::SHA2;
use whir::protocols::{
    irs_commit, matrix_commit, merkle_tree, proof_of_work, sumcheck,
    whir as whir_mod,
};
use whir::transcript::{
    codecs::Empty, DomainSeparator, ProverState, VerifierMessage, VerifierState,
};

type F = Field64;

fn dump_bytes(label: &str, data: &[u8]) {
    print!("    {label} ");
    for byte in data { print!("{byte:02x}"); }
    println!();
}

fn dump_field_vec(label: &str, vec: &[F]) {
    print!("    {label}");
    for v in vec { print!(" {v}"); }
    println!();
}

struct Lcg(u64);
impl Lcg {
    fn new(seed: u64) -> Self { Self(seed) }
    fn next_u64(&mut self) -> u64 {
        self.0 = self.0.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        self.0
    }
    fn next_field64(&mut self) -> F { F::from(self.next_u64()) }
}

/// 为每个 case 构造独立的 DomainSeparator (session 不同, 区分海绵状态)
fn make_ds(label: &str) -> DomainSeparator<'static, Empty> {
    let session_str = format!("whir_{label}");
    DomainSeparator::protocol(&0xBEEFu32)
        .session(&session_str)
        .instance(&Empty)
}

/// 运行单个 WHIR 测试 case
fn run_case(
    case_id: usize,
    num_vectors: usize,
    vector_size: usize,
    rng: &mut Lcg,
) {
    let depth: usize = 1;
    let cw = vector_size * 2;
    let num_cols = num_vectors * depth;
    let ood = 1usize;
    let ind = 3.min(vector_size);
    let log_vs = vector_size.ilog2() as usize;

    println!("CASE {case_id}");

    // ---- 1. 输入构造 ----
    // 每个多项式生成随机系数
    let mut vecs: Vec<Vec<F>> = Vec::with_capacity(num_vectors);
    let mut evals = Vec::with_capacity(num_vectors);
    for vi in 0..num_vectors {
        let v: Vec<F> = (0..vector_size).map(|_| rng.next_field64()).collect();
        dump_field_vec(&format!("input_vector{vi}"), &v);
        evals.push(v.iter().sum());  // 全 1 线性形式在每向量上的求值 = sum(coeffs)
        vecs.push(v);
    }
    let vec_slices: Vec<&[F]> = vecs.iter().map(|v| v.as_slice()).collect();

    // 一个全 1 线性形式, 对每个多项式做一个求值
    let cv = (0..vector_size).map(|_| F::ONE).collect::<Vec<_>>();
    let lfs: Vec<Box<dyn LinearForm<F>>> =
        vec![Box::new(Covector { vector: cv })];

    // ---- 2. 协议配置 ----
    let mt = merkle_tree::Config::with_hash(SHA2, cw);
    let irs = irs_commit::Config::<Identity<F>> {
        embedding: Default::default(),
        num_vectors,
        vector_size,
        codeword_length: cw,
        interleaving_depth: depth,
        matrix_commit: matrix_commit::Config::<F> {
            element_type: Default::default(),
            num_cols,
            leaf_hash_id: SHA2,
            merkle_tree: mt,
        },
        johnson_slack: OrderedFloat(0.0),
        in_domain_samples: ind,
        out_domain_samples: ood,
        deduplicate_in_domain: false,
    };

    let sc_init = sumcheck::Config::<F> {
        field: Default::default(),
        initial_size: vector_size,
        round_pow: proof_of_work::Config { hash_id: SHA2, threshold: u64::MAX },
        num_rounds: 0,
    };
    let skip_pow = proof_of_work::Config { hash_id: SHA2, threshold: u64::MAX };
    let sc_final = sumcheck::Config::<F> {
        field: Default::default(),
        initial_size: vector_size,
        round_pow: proof_of_work::Config { hash_id: SHA2, threshold: u64::MAX },
        num_rounds: log_vs,
    };

    let whir_config = whir_mod::Config::<Identity<F>> {
        initial_committer: irs,
        initial_sumcheck: sc_init,
        initial_skip_pow: skip_pow,
        round_configs: vec![],
        final_sumcheck: sc_final,
        final_pow: proof_of_work::Config { hash_id: SHA2, threshold: u64::MAX },
    };

    // ---- 3. DomainSeparator ----
    let ds = make_ds(&case_id.to_string());

    // ---- 4. Prover ----
    let mut ps = ProverState::new_std(&ds);
    let witness = whir_config.commit(&mut ps, &vec_slices);

    dump_field_vec("witness_matrix", &witness.matrix);
    dump_field_vec("ood_points", &witness.out_of_domain.points);
    dump_field_vec("ood_matrix", &witness.out_of_domain.matrix);

    let witnesses: Vec<Cow<irs_commit::Witness<F, F>>> = vec![Cow::Owned(witness)];
    let vec_cows: Vec<Cow<[F]>> = vec_slices.iter().map(|v| Cow::Borrowed(*v)).collect();

    let claim = whir_config.prove(
        &mut ps, vec_cows, witnesses,
        lfs, Cow::Borrowed(&evals),
    );
    let proof = ps.proof();

    dump_field_vec("eval_point", &claim.evaluation_point);
    dump_field_vec("rlc_coeffs", &claim.rlc_coefficients);
    dump_bytes("proof_narg", &proof.narg_string);
    dump_bytes("proof_hints", &proof.hints);

    // ---- 5. Verifier ----
    let mut vs = VerifierState::new_std(&ds, &proof);
    let comm = whir_config.receive_commitment(&mut vs).unwrap();
    let vc = whir_config.verify(&mut vs, &[&comm], &evals).unwrap();

    dump_field_vec("v_eval_point", &vc.evaluation_point);
    dump_field_vec("v_rlc_coeffs", &vc.rlc_coefficients);
    println!("    check_eof {}", vs.check_eof().is_ok() as i32);
}

fn main() {
    let mut rng = Lcg::new(0xCAFE);
    println!("# SECTION whir");

    // CASE 0: 1 多项式 × 2 系数 (最小)
    run_case(0, 1, 2, &mut rng);

    // CASE 1: 1 多项式 × 4 系数
    run_case(1, 1, 4, &mut rng);

    // CASE 2: 1 多项式 × 8 系数
    run_case(2, 1, 8, &mut rng);

    // CASE 3: 3 多项式 × 4 系数 (多向量)
    run_case(3, 3, 4, &mut rng);

    // CASE 4: 4 多项式 × 8 系数 (多向量 + 大参数)
    run_case(4, 4, 8, &mut rng);
}
